vic/drivers/shared_image/src/vic_write.c

```diff
@@ -27,114 +27,221 @@
+    extern nc_file_struct *nc_hist_files;
+
+    size_t                 stream_idx;
+
+    // Write data
+    for (stream_idx = 0; stream_idx < options.Noutstreams; stream_idx++) {
+        if (raise_alarm(&(output_streams[stream_idx].agg_alarm), dmy)) {
+            debug("raised alarm for stream %zu", stream_idx);
+            vic_write(&(output_streams[stream_idx]),
+                      &(nc_hist_files[stream_idx]), dmy);
+            reset_stream(&(output_streams[stream_idx]), dmy);
+        }
+    }
+}
+
+/******************************************************************************
  * @brief    Write output to netcdf file. Currently everything is cast to
  *           double
  *****************************************************************************/
 void
-vic_write(dmy_struct *dmy_current)
+vic_write(stream_struct  *stream,
+          nc_file_struct *nc_hist_file,
+          dmy_struct     *dmy_current)
 {
-    extern out_data_struct   **out_data;
     extern global_param_struct global_param;
     extern domain_struct       local_domain;
-    extern nc_file_struct      nc_hist_file;
-    extern nc_var_struct       nc_vars[N_OUTVAR_TYPES];
-    extern size_t              current;
     extern int                 mpi_rank;
-    int                        dimids[MAXDIMS];
+    extern metadata_struct     out_metadata[N_OUTVAR_TYPES];
+
     size_t                     i;
     size_t                     j;
     size_t                     k;
     size_t                     ndims;
+    double                     dtime;
     double                    *dvar = NULL;
+    float                     *fvar = NULL;
+    int                       *ivar = NULL;
+    short int                 *svar = NULL;
+    char                      *cvar = NULL;
     size_t                     dcount[MAXDIMS];
     size_t                     dstart[MAXDIMS];
+    unsigned int               varid;
+    int                        status;
 
 
-    // allocate memory for variables to be stored
-    dvar = malloc(local_domain.ncells_active * sizeof(*dvar));
-    if (dvar == NULL) {
-        log_err("Memory allocation error in vic_write().");
-    }
-
-    // set missing values
-    for (i = 0; i < local_domain.ncells_active; i++) {
-        dvar[i] = nc_hist_file.d_fillvalue;
-    }
-
     // initialize dimids to invalid values - helps debugging
     for (i = 0; i < MAXDIMS; i++) {
-        dimids[i] = -1;
         dstart[i] = -1;
-        dcount[i] = -1;
+        dcount[i] = 0;
     }
 
-    for (k = 0; k < N_OUTVAR_TYPES; k++) {
-        if (!nc_vars[k].nc_write) {
-            continue;
+    for (k = 0; k < stream->nvars; k++) {
+        varid = stream->varid[k];
+
+        if (nc_hist_file->nc_vars[k].nc_type == NC_DOUBLE) {
+            if (dvar == NULL) {
+                // allocate memory for variables to be stored
+                dvar = malloc(local_domain.ncells_active * sizeof(*dvar));
+                if (dvar == NULL) {
```

Comment:
We should make a `check_alloc_status()` macro or function the same way we do for `nc_check_status()`. Leave as is and open a separate issue if we do not already have one.